File: tags/release-0.81/src/gameopt.cpp
// File:     gameopt.cpp
// Location: src
// Purpose:  implement GameOpt class
// Author:   Stephen Gold sgold@sonic.net
// (c) Copyright 2012 Stephen Gold
// Distributed under the terms of the GNU General Public License

/*
This file is part of the Gold Tile Game.

The Gold Tile Game is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the 
Free Software Foundation, either version 3 of the License, or (at your 
option) any later version.

The Gold Tile Game is distributed in the hope that it will be useful, but 
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You should have received a copy of the GNU General Public License
along with the Gold Tile Game.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <iostream>
#include "gameopt.hpp"


// lifecycle

GameOpt::GameOpt(void) {
	mMinutesPerHand = MINUTES_PER_HAND_DEFAULT;
	mStyle = GAME_STYLE_DEFAULT;
	Standardize();
	Validate();
}

// The compiler-generated copy constructor is OK.
// The compiler-generated destructor is OK.


// operators

// The compiler-generated assignment method is OK.

GameOpt::operator GameStyleType(void) const {
	ASSERT(mStyle != GAME_STYLE_NONE);

	return mStyle;
}

GameOpt::operator GridType(void) const {
	return mGrid;
}

GameOpt::operator RulesType(void) const {
	return mRules;
}


// misc methods

AttrCntType GameOpt::AttrCnt(void) const {
	ASSERT(mAttrCnt >= Combo::ATTRIBUTE_CNT_MIN);
#ifdef _GUI
	ASSERT(mAttrCnt <= Combo::ATTRIBUTE_CNT_MAX);
#endif // defined(_GUI)

	return mAttrCnt;
}

String GameOpt::AttrReport(void) const {
	String result = ::plural(AttrCnt(), "attribute") + ":\n";
	for (AttrIndexType i_attr = 0; i_attr < AttrCnt(); i_attr++) {
		AttrType const value_max = MaxAttrValue(i_attr);
		result += " " + ::ordinal(i_attr + 1) + " attribute ranges from ";

		AttrModeType const display_mode = Combo::DefaultDisplayMode(i_attr);
		result += Combo::AttributeToString(display_mode, 0) + " to ";
		result += Combo::AttributeToString(display_mode, value_max) + "\n";
	}
	result += "\n";

	return result;
}

RowType GameOpt::BoardHeight(void) const {
	ASSERT(::is_even(mBoardHeight));
	ASSERT(mBoardHeight >= Cell::HEIGHT_MIN);
	ASSERT(mBoardHeight <= Cell::HEIGHT_MAX);

	return mBoardHeight;
}

ColumnType GameOpt::BoardWidth(void) const {
	ASSERT(::is_even(mBoardWidth));
	ASSERT(mBoardWidth >= Cell::WIDTH_MIN);
	ASSERT(mBoardWidth <= Cell::WIDTH_MAX);

	return mBoardWidth;
}

unsigned GameOpt::BonusPercent(void) const {
	ASSERT(mBonusPercent < 100);

	return mBonusPercent;
}

long GameOpt::CloneCnt(void) const {
	long const result = ComboCnt() * ClonesPerCombo();

	return result;
}

unsigned GameOpt::ClonesPerCombo(void) const {
	return mClonesPerCombo;
}

long GameOpt::ComboCnt(void) const {
	// count the possible combinations
	long result = 1L;
	for (AttrIndexType i_attr = 0; i_attr < mAttrCnt; i_attr++) {
	    AttrType const possible_values = CountAttrValues(i_attr);
		result *= possible_values;
	}
	ASSERT(result >= Combo::COMBINATION_CNT_MIN);
#ifdef _GUI
    ASSERT(result <= Combo::COMBINATION_CNT_MAX);
#endif // defined(_GUI)

	return result;
}

AttrType GameOpt::CountAttrValues(AttrIndexType ind) const {
	AttrType result = Combo::VALUE_CNT_DEFAULT;

	if (ind < mMaxAttrValues.size()) {
		result = mMaxAttrValues[ind] + 1;  // including zero
	}

	ASSERT(result >= Combo::VALUE_CNT_MIN);
	ASSERT(result <= Combo::VALUE_CNT_MAX);
	return result;
}

void GameOpt::GetUserChoice(void) {
	String const attr_report = AttrReport();
	std::cout << attr_report;
	
	int hands_dealt = 0;
	while (hands_dealt < int(HANDS_DEALT_MIN)) {
	    std::cout << "Deal how many hands? ";
		String line;
	    std::cin >> line;
		hands_dealt = int(line);
	}
	SetHandsDealt(hands_dealt);
	std::cout << std::endl;

	int hand_size = 0;
	while (hand_size < int(HAND_SIZE_MIN)) {
	    std::cout << "How many tiles per hand? ";
		String line;
	    std::cin >> line;
		hand_size = int(line);
	}
	SetHandSize(hand_size);
	std::cout << std::endl;
	
	// Clone tiles so that there are enough to fill each hand at least three times.
	unsigned const tiles_needed = 3 * hand_size * hands_dealt;
	long const combo_cnt = ComboCnt();
	unsigned clones_per_combo = unsigned(tiles_needed / combo_cnt);
	SetClonesPerCombo(clones_per_combo);
}

unsigned GameOpt::HandsDealt(void) const {
	ASSERT(mHandsDealt >= HANDS_DEALT_MIN);

	return mHandsDealt;
}

unsigned GameOpt::HandSize(void) const {
	ASSERT(mHandSize >= HAND_SIZE_MIN);

	return mHandSize;
}

AttrType GameOpt::MaxAttrValue(AttrIndexType ind) const {
	AttrType const result = CountAttrValues(ind) - 1; // excluding zero

	return result;
}

unsigned GameOpt::MinutesPerHand(void) const {
	ASSERT(mMinutesPerHand >= MINUTES_PER_HAND_MIN);
	
	return mMinutesPerHand;
}

unsigned GameOpt::SecondsPerHand(void) const {
	unsigned const result = MinutesPerHand() * SECONDS_PER_MINUTE;

	return result;
}

void GameOpt::SetAttrCnt(AttrCntType attrCnt) {
	ASSERT(attrCnt >= Combo::ATTRIBUTE_CNT_MIN);
#ifdef _GUI
	ASSERT(attrCnt <= Combo::ATTRIBUTE_CNT_MAX);
#endif // defined(_GUI)

	mAttrCnt = attrCnt;
}

void GameOpt::SetBoardHeight(RowType height) {
	ASSERT(::is_even(height));
	ASSERT(height >= Cell::HEIGHT_MIN);
	ASSERT(height <= Cell::HEIGHT_MAX);

	mBoardHeight = height;
}

void GameOpt::SetBoardWidth(ColumnType width) {
	ASSERT(::is_even(width));
	ASSERT(width >= Cell::WIDTH_MIN);
	ASSERT(width <= Cell::WIDTH_MAX);

	mBoardWidth = width;
}

void GameOpt::SetBonusPercent(unsigned percent) {
	ASSERT(percent < 100);

	mBonusPercent = percent;
}

void GameOpt::SetChallenge(void) {
	mStyle = GAME_STYLE_CHALLENGE;
}

void GameOpt::SetClonesPerCombo(unsigned clonesPerCombo) {
	mClonesPerCombo = clonesPerCombo;
}

void GameOpt::SetDebug(void) {
	mStyle = GAME_STYLE_DEBUG;
}

void GameOpt::SetDoesBoardWrap(bool wrap) {
	mDoesBoardWrap = wrap;
}

void GameOpt::SetFriendly(void) {
	mStyle = GAME_STYLE_FRIENDLY;
}

void GameOpt::SetGrid(GridType grid) {
	mGrid = grid;
}

void GameOpt::SetHandsDealt(unsigned handsDealt) {
	ASSERT(mHandsDealt >= HANDS_DEALT_MIN);

	mHandsDealt = handsDealt;
}

void GameOpt::SetHandSize(unsigned handSize) {
	ASSERT(handSize >= HAND_SIZE_MIN);

	mHandSize = handSize;
}

void GameOpt::SetMinutesPerHand(unsigned minutes) {
	ASSERT(minutes >= MINUTES_PER_HAND_MIN);

	mMinutesPerHand = minutes;
}

void GameOpt::SetNumAttrValues(AttrIndexType ind, AttrType numValues) {
	ASSERT(numValues >= Combo::VALUE_CNT_MIN);
	ASSERT(numValues <= Combo::VALUE_CNT_MAX);

	AttrCntType const original_count = AttrCntType(mMaxAttrValues.size());
	AttrType const max_value = Combo::VALUE_CNT_DEFAULT - 1; // excluding zero
	for (AttrIndexType i_attr = original_count; i_attr <= ind; i_attr++) {
        mMaxAttrValues.push_back(max_value);
	}
	ASSERT(mMaxAttrValues.size() > ind);
	mMaxAttrValues[ind] = numValues - 1;  // excluding zero
}

void GameOpt::SetPractice(void) {
	mStyle = GAME_STYLE_PRACTICE;
}

void GameOpt::SetRules(RulesType rules) {
	mRules = rules;
}

void GameOpt::Standardize(void) {
    mAttrCnt = Combo::ATTRIBUTE_CNT_DEFAULT;
	mMaxAttrValues.clear();
	for (unsigned i_attr = 0; i_attr < mAttrCnt; i_attr++) {
	    mMaxAttrValues.push_back(Combo::VALUE_CNT_DEFAULT - 1);
	}

    mBoardHeight = Cell::HEIGHT_MAX;
	mBoardWidth = Cell::WIDTH_MAX;
	mBonusPercent = BONUS_PERCENT_DEFAULT;
	mClonesPerCombo = CLONES_PER_COMBO_DEFAULT;
	mDoesBoardWrap = false;
	mGrid = GRID_DEFAULT;
	mHandsDealt = HAND_CNT_DEFAULT;
	mHandSize = HAND_SIZE_DEFAULT;
	mRules = RULES_STANDARD;
}

void GameOpt::StyleChange(void) {
	switch (mStyle) {
		case GAME_STYLE_PRACTICE:
		    mHandsDealt = 1;
			break;

		case GAME_STYLE_CHALLENGE:
	    case GAME_STYLE_DEBUG:
		case GAME_STYLE_FRIENDLY:
			break;

		default:
			FAIL();
	}
}

unsigned GameOpt::TilesPerCombo(void) const {
	unsigned const result = ClonesPerCombo() + 1;

	return result;
}

long GameOpt::TotalTileCnt(void) const {
	long const result = ComboCnt() * (1 + ClonesPerCombo());

	return result;
}

void GameOpt::Validate(void) const {
	ASSERT(mAttrCnt >= Combo::ATTRIBUTE_CNT_MIN);
#ifdef _GUI
	ASSERT(mAttrCnt <= Combo::ATTRIBUTE_CNT_MAX);
#endif // defined(_GUI)
	ASSERT(::is_even(mBoardHeight));
	ASSERT(mBoardHeight >= Cell::HEIGHT_MIN);
	ASSERT(mBoardHeight <= Cell::HEIGHT_MAX);
	ASSERT(::is_even(mBoardWidth));
	ASSERT(mBoardWidth >= Cell::WIDTH_MIN);
	ASSERT(mBoardWidth <= Cell::WIDTH_MAX);
	ASSERT(mBonusPercent < 100);
	ASSERT(mHandsDealt >= HANDS_DEALT_MIN);
	ASSERT(mHandSize >= HAND_SIZE_MIN);
	ASSERT(mMinutesPerHand >= MINUTES_PER_HAND_MIN);
	ASSERT(mStyle != GAME_STYLE_NONE);
}


// inquiry methods

bool GameOpt::DoesBoardWrap(void) const {
	return mDoesBoardWrap;
}

bool GameOpt::HasFiniteHeight(void) const {
	bool const result = (BoardHeight() < Cell::HEIGHT_MAX);

	return result;
}

bool GameOpt::HasFiniteWidth(void) const {
	bool const result = (BoardWidth() < Cell::WIDTH_MAX);

	return result;
}

bool GameOpt::HasTimeLimit(void) const {
	bool const result = IsChallenge();

	return result;
}

bool GameOpt::IsChallenge(void) const {
	bool const result = (mStyle == GAME_STYLE_CHALLENGE);

	return result;
}

bool GameOpt::IsDebug(void) const {
	bool const result = (mStyle == GAME_STYLE_DEBUG);

	return result;
}

bool GameOpt::IsFriendly(void) const {
	bool const result = (mStyle == GAME_STYLE_FRIENDLY);

	return result;
}

bool GameOpt::IsPractice(void) const {
	bool const result = (mStyle == GAME_STYLE_PRACTICE);

	return result;
}

File: Java/GoldTile/src/goldtile/GameOpt.java
// File:     GameOpt.java
// Location: Java/GoldTile/src/goldtile
// Purpose:  GameOpt class for the Gold Tile Game
/**
 * @author Stephen Gold
 */
// (c) Copyright 2012 Stephen Gold
// Distributed under the terms of the GNU General Public License

/*
This file is part of the Gold Tile Game.

The Gold Tile Game is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

The Gold Tile Game is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with the Gold Tile Game.  If not, see <http://www.gnu.org/licenses/>.
*/

package goldtile;

import java.util.Date;

public class GameOpt
    implements ReadGameOpt
{
    // constants, sorted by type
    final public static Dim BOARD_HEIGHT_DEFAULT = new Dim(12);
    final public static Dim BOARD_WIDTH_DEFAULT = BOARD_HEIGHT_DEFAULT;
    final public static Fraction BONUS_FRACTION_DEFAULT = new Fraction(0.0);
    final public static Fraction BONUS_FRACTION_MIN = new Fraction(0.0);
    final public static int ATTR_COUNT_DEFAULT = 2;
    final public static int ATTR_COUNT_MAX = 5;
    final public static int ATTR_COUNT_MIN = 2;
    final public static int CLONES_PER_COMBO_DEFAULT = 2;
    final public static int CLONES_PER_COMBO_MIN = 0;
    final public static int HANDS_DEALT_DEFAULT = 2;
    final public static int HANDS_DEALT_MIN = 1;
    final public static int HAND_SIZE_DEFAULT = 6;
    final public static int HAND_SIZE_MIN = 1;
    final public static int MINUTES_PER_HAND_DEFAULT = 30;
    final public static int MINUTES_PER_HAND_MAX = 120;
    final public static int MINUTES_PER_HAND_MIN = 2;
    final private static int STUCK_THRESHOLD_DEFAULT = 7;
    final private static int STUCK_THRESHOLD_MIN = 1;
    final public static long SEED_DEFAULT = 12345L;

    // per-instance fields, sorted by type
    final private Attr lastAttrs[] = new Attr[ATTR_COUNT_MAX];
                                             // maximum value of each attribute
    private boolean boardWrap;
                      // used only if board is finite on one or both dimensions
    private boolean randomize;
    private Dim boardHeight;
    private Dim boardWidth;
    private Fraction bonusFraction;  // fraction of tiles with bonus value
    private GameStyle style;
    private Grid grid;
    private int attrCount;
    private int clonesPerCombo;
    private int handsDealt;      // number of hands dealt
    private int handSize;        // maximum number of tiles in a hand
    private int minutesPerHand;  // used only in challenge-style games
    private int stuckThreshold;  // turns before game is declared "stuck"
    private long seed;  // used only in debug-style with randomizeFlag == false
    private Rules rules;

    // constructors

    public GameOpt() {
        minutesPerHand = MINUTES_PER_HAND_DEFAULT;
        randomize = true;
        seed = SEED_DEFAULT;
        style = GameStyle.getDefault();

        standardize();
        validate();
    }

    public GameOpt(ReadGameOpt other) {
        assert other != null;

        boardWrap = other.doesBoardWrap();
        randomize = other.isRandomized();
        style = other.getStyle();
        grid = other.getGrid();
        attrCount = other.getAttrCount();
        boardHeight = other.getBoardHeight();
        boardWidth = other.getBoardWidth();
        clonesPerCombo = other.getClonesPerCombo();
        handsDealt = other.getHandsDealt();
        handSize = other.getHandSize();
        for (int iAttr = 0; iAttr < ATTR_COUNT_MAX; iAttr++) {
            lastAttrs[iAttr] = other.getAttrLast(iAttr);
        }
        minutesPerHand = other.getMinutesPerHand();
        stuckThreshold = other.getStuckThreshold();
        seed = other.getSeed();
        bonusFraction = other.getBonusFraction();
        rules = other.getRules();

        validate();
    }

    // methods

    public static GameOpt chooseConsole() {
        final GameOpt result = new GameOpt();

        result.setRandomize(false);

        final String attrReport = result.reportAttrs();
        Console.print(attrReport);

        result.handsDealt = -1;
        while (result.handsDealt < HANDS_DEALT_MIN) {
            String line = Console.readLine("Deal how many hands? ");
            try {
                result.handsDealt = Integer.parseInt(line);
            } catch (NumberFormatException exception) {
            }
        }
        Console.printLine();

        result.handSize = -1;
        while (result.handSize < HAND_SIZE_MIN) {
            String line = Console.readLine("How many tiles per hand? ");
            try {
                result.handSize = Integer.parseInt(line);
            } catch (NumberFormatException exception) {
            }
        }
        Console.printLine();

        /*
         * Clone tiles so that there are enough to fill every hand at
         * least three times.
         */
        final long tilesNeeded = 3 * result.handSize * result.handsDealt;
        final long comboCount = result.countCombos();
        result.clonesPerCombo = (int)(tilesNeeded / comboCount);

        return result;
    }

    @Override
    public long countCombos() {
        long result = 1;
        for (int iAttr = 0; iAttr < attrCount; iAttr++) {
            result *= getAttrValueCount(iAttr);
        }

        return result;
    }

    @Override
    public long countTiles() {
        return countCombos() * (1 + clonesPerCombo);
    }

    @Override
    public boolean doesBoardWrap() {
        return boardWrap;
    }

    @Override
    public int getAttrCount() {
        return attrCount;
    }

    @Override
    public Attr getAttrLast(int iAttr) {
        assert iAttr >= 0;
        assert iAttr < lastAttrs.length : iAttr;

        return lastAttrs[iAttr];
    }

    @Override
    public int getAttrLastValue(int iAttr) {
        assert iAttr >= 0;
        assert iAttr < lastAttrs.length : iAttr;

        return  getAttrLast(iAttr).intValue();
    }

    @Override
    public int getAttrValueCount(int iAttr) {
        assert iAttr >= 0;
        assert iAttr < lastAttrs.length : iAttr;

        return 1 + getAttrLastValue(iAttr);
    }

    @Override
    public Dim getBoardHeight() {
        return boardHeight;
    }

    @Override
    public Dim getBoardWidth() {
        return boardWidth;
    }

    @Override
    public Fraction getBonusFraction() {
        return bonusFraction;
    }

    @Override
    public long getCellCount() {
        final long product = boardWidth.times(boardHeight);
        if (product == Long.MAX_VALUE) {
            return product;
        } else {
            return product / grid.getSparsity();
        }
    }

    @Override
    public int getClonesPerCombo() {
        return clonesPerCombo;
    }

    @Override
    public Grid getGrid() {
        return grid;
    }

    @Override
    public int getHandsDealt() {
        return handsDealt;
    }

    @Override
    public int getHandSize() {
        return handSize;
    }

    @Override
    public long getMillisecondsPerHand() {
        return getSecondsPerHand() * Global.MILLISECONDS_PER_SECOND;
    }

    @Override
    public int getMinutesPerHand() {
        return minutesPerHand;
    }

    @Override
    public Rules getRules() {
        return rules;
    }

    @Override
    public int getSecondsPerHand() {
        return getMinutesPerHand() * Global.SECONDS_PER_MINUTE;
    }

    @Override
    public long getSeed() {
        return seed;
    }

    @Override
    public int getStuckThreshold() {
        return stuckThreshold;
    }

    @Override
    public GameStyle getStyle() {
        return style;
    }

    @Override
    public int getTilesPerCombo() {
        return 1 + getClonesPerCombo();
    }

    @Override
    public Topology getTopology() {
        if (hasFiniteHeight() && hasFiniteWidth()) {
            return doesBoardWrap() ? Topology.TORUS : Topology.RECTANGLE;
        } else if (hasFiniteHeight()) {
            return doesBoardWrap() ? Topology.HCYLINDER : Topology.HSTRIP;
        } else if (hasFiniteWidth()) {
            return doesBoardWrap() ? Topology.VCYLINDER : Topology.VSTRIP;
        } else {
            return Topology.FLAT;
        }
    }

    @Override
    public boolean hasFiniteHeight() {
        return !boardHeight.isEndless();
    }

    @Override
    public boolean hasFiniteWidth() {
        return !boardWidth.isEndless();
    }

    @Override
    public boolean isRandomized() {
        return randomize;
    }

    @Override
    public String reportAttrs() {
        String result = "Tiles have ";
        result += StringExt.plural(attrCount, "attribute") + ":\n";

        for (int iAttr = 0; iAttr < attrCount; iAttr++) {
            final AttrMode mode = AttrMode.getConsoleDefault(iAttr);

            result += String.format(" The %s attribute ranges from %s to %s.\n",
                    StringExt.ordinal(iAttr + 1),
                    mode.attrToString(new Attr(0)),
                    mode.attrToString(getAttrLast(iAttr)) );
        }
        result += "\n";

        return result;
    }

    public void reseedGenerator() {
        if (randomize || !style.isDebug()) {
            final Date date = new Date();
            seed = date.getTime();
        }
        Global.reseedGenerator(seed);
    }

    public void setAttrCount(int attrCount) {
        assert attrCount >= ATTR_COUNT_MIN;
        assert attrCount <= ATTR_COUNT_MAX;

        this.attrCount = attrCount;
    }

    public void setAttrValueCount(int iAttr, int count) {
       assert iAttr >= 0;
       assert iAttr < ATTR_COUNT_MAX;
       assert count >= Attr.COUNT_MIN;
       assert count <= Attr.COUNT_MAX;

       lastAttrs[iAttr] = new Attr(count - 1);
    }

    public void setBoardHeight(Dim boardHeight) {
        this.boardHeight = boardHeight;
    }

    public void setBoardWidth(Dim boardWidth) {
        this.boardWidth = boardWidth;
    }

    public void setBoardWrap(boolean doesWrap) {
        this.boardWrap = doesWrap;
    }

    public void setBonusFraction(Fraction bonusFraction) {
        assert bonusFraction != null;

        this.bonusFraction = bonusFraction;
    }

    public void setClonesPerCombo(int clonesPerCombo) {
        assert clonesPerCombo >= CLONES_PER_COMBO_MIN;

        this.clonesPerCombo = clonesPerCombo;
    }

    public void setGrid(Grid grid) {
        assert grid != null;

        this.grid = grid;
    }

    public void setHandSize(int handSize) {
        assert handSize >= HAND_SIZE_MIN;

        this.handSize = handSize;
    }

    public void setHandsDealt(int handsDealt) {
        assert handsDealt >= HANDS_DEALT_MIN;

        this.handsDealt = handsDealt;
    }

    public void setMinutesPerHand(int minutesPerHand) {
        this.minutesPerHand = minutesPerHand;
    }

    public void setRandomize(boolean randomizeFlag) {
        this.randomize = randomizeFlag;
    }

    public void setRules(Rules rules) {
        assert rules != null;

        this.rules = rules;
    }

    public void setSeed(long seed) {
        this.seed = seed;
    }

    public void setStyle(GameStyle style) {
        assert style != null;

        this.style = style;
    }

    final public void standardize() {
        attrCount = ATTR_COUNT_DEFAULT;
        for (int iAttr = 0; iAttr < ATTR_COUNT_MAX; iAttr++) {
            lastAttrs[iAttr] = new Attr(Attr.LAST_DEFAULT);
        }
        boardHeight = Dim.endless();
        boardWidth = Dim.endless();
        bonusFraction = BONUS_FRACTION_DEFAULT;
        clonesPerCombo = CLONES_PER_COMBO_DEFAULT;
        boardWrap = false;
        grid = Grid.getDefault();
        handsDealt = HANDS_DEALT_DEFAULT;
        handSize = HAND_SIZE_DEFAULT;
        // don't set minutesPerHand
        stuckThreshold = STUCK_THRESHOLD_DEFAULT;
        // don't set randomizeFlag
        rules = Rules.STANDARD;
        // don't set seed
        // don't set style
    }

    @Override
    final public void validate() {
        assert attrCount <= ATTR_COUNT_MAX : attrCount;
        assert attrCount >= ATTR_COUNT_MIN : attrCount;
        for (int iAttr = 0; iAttr < ATTR_COUNT_MAX; iAttr++) {
            assert lastAttrs[iAttr].intValue() >= Attr.LAST_MIN;
            assert lastAttrs[iAttr].intValue() <= Attr.LAST_MAX;
        }
        assert boardHeight != null;
        assert boardWidth != null;
        assert bonusFraction != null;
        assert clonesPerCombo >= CLONES_PER_COMBO_MIN : clonesPerCombo;
        assert grid != null;
        assert handsDealt >= HANDS_DEALT_MIN : handsDealt;
        assert handSize >= HAND_SIZE_MIN : handSize;
        assert lastAttrs != null;
        assert lastAttrs.length >= attrCount : lastAttrs;
        assert minutesPerHand >= MINUTES_PER_HAND_MIN : minutesPerHand;
        assert stuckThreshold >= STUCK_THRESHOLD_MIN : stuckThreshold;
        assert rules != null;
        assert style != null;
        assert style != GameStyle.NONE;
    }
}
